x/incentives/keeper/grpc_query.go

```diff
@@ -269,6 +270,91 @@ func (q Querier) getGaugeFromIDJsonBytes(ctx sdk.Context, refValue []byte) ([]ty
 }
 
+// GaugeFilterFn is used to apply a filter on gauges
+// It must be used in query.FilterPaginate as a condition to add the gauge to the response data
+type GaugesFilterFn func(gauge *types.Gauge) bool
+
+func (q Querier) GaugesByPoolID(goCtx context.Context, req *types.QueryGaugesByPoolIDRequest) (*types.QueryGaugesByPoolIDResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pageRes, gauges, err := q.getGaugesWithFilter(ctx, types.KeyPrefixGauges, req.GetPagination(), func(gauge *types.Gauge) bool {
+		return gauge.IsLinkedToPool(req.GetId())
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &types.QueryGaugesByPoolIDResponse{
+		Gauges:     gauges,
+		Pagination: pageRes,
+	}, nil
+}
+
+func (q Querier) InternalGauges(goCtx context.Context, req *types.QueryInternalGaugesRequest) (*types.QueryInternalGaugesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pageRes, gauges, err := q.getGaugesWithFilter(ctx, types.KeyPrefixGauges, req.GetPagination(), func(gauge *types.Gauge) bool {
+		return gauge.IsInternalGauge([]string{gammtypes.GAMMTokenPrefix})
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &types.QueryInternalGaugesResponse{
+		Gauges:     gauges,
+		Pagination: pageRes,
+	}, nil
+}
+
+func (q Querier) ExternalGauges(goCtx context.Context, req *types.QueryExternalGaugesRequest) (*types.QueryExternalGaugesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pageRes, gauges, err := q.getGaugesWithFilter(ctx, types.KeyPrefixGauges, req.GetPagination(), func(gauge *types.Gauge) bool {
+		return gauge.IsExternalGauge()
+	})
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &types.QueryExternalGaugesResponse{
+		Gauges:     gauges,
+		Pagination: pageRes,
+	}, nil
+}
+
+func (q Querier) getGaugesWithFilter(ctx sdk.Context, keyPrefix []byte, pageReq *query.PageRequest, filter GaugesFilterFn) (*query.PageResponse, []types.Gauge, error) {
+	gauges := []types.Gauge{}
+	store := ctx.KVStore(q.Keeper.storeKey)
+	valStore := prefix.NewStore(store, keyPrefix)
+
+	pageRes, err := query.FilteredPaginate(valStore, pageReq, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		// this may return multiple gauges at once if two gauges start at the same time.
+		// for now this is treated as an edge case that is not of importance
+		newGauges, err := q.getGaugeFromIDJsonBytes(ctx, value)
+		if err != nil {
+			return false, err
+		}
+		var hit bool
+		if accumulate {
+			for _, gauge := range newGauges {
+				if !filter(&gauge) {
+					continue
+				}
+				gauges = append(gauges, gauge)
+				hit = true
+			}
+		}
+		return hit, nil
+	})
```

Comment:
@mattverse 
by the way the request could have some edge case where number of items is greater than the limit.
As explained by the comment ( the same than for the existing fuction `filterByPrefixAndDenom`)
it is because of storing a key->value or key->[1]value in some situation more than one value is stored for the given key but it looks like it is rare

Reply:
Can you add a comment so we don't get confused in the future?

Reply:
Comment is already added 🙂 I copied pasted the one from filterByPrefixAndDenom